### PR TITLE
Add a compute-only MAC/AC energy estimator with Horowitz presets

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary
- Add a compute-only MAC/AC energy profiler with both context-manager and convenience function APIs.
- Reuse existing `MACCounter`, `ACCounter`, `SynOpCounter`, and `FlopCounter` to expose normalized compute-energy estimates.
- Add Horowitz-style FP32/FP16/INT8 cost presets and keep FP32 as the default comparison regime.
- Update op_counter exports and API docs to surface the new estimator.
- Include unit tests covering the new profiler, presets, and default-cost behavior.

## Testing
- Not run (not requested).

## Summary by CodeRabbit

* **New Features**
  * Added a compute-only energy estimator with configurable cost presets, a profiler, and a convenience API that returns structured energy reports.

* **Documentation**
  * Added a "Compute‑Only Energy Estimator" API doc section and expanded bilingual (Chinese/English) API docstrings across counters and utilities; fixed typos and parameter annotations.

* **Tests**
  * Added tests covering estimation behavior, presets, reporting consistency, input forms, and strict vs. warning modes.

* **Chores**
  * Made the energy estimator available at the package level and adjusted a counter initializer type annotation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/fangwei123456/spikingjelly/pull/681?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Approved ✅** (3 suggestions)

**PR:** #681 — Add a compute-only MAC/AC energy estimator with Horowitz presets
**Author:** @AllenYolk
**Files changed:** 13 (+1029 −27)

---

### ✅ Looks Good

**Clean composition pattern** — The new `ComputeEnergyProfiler` elegantly composes existing `MACCounter`, `ACCounter`, `SynOpCounter`, and `FlopCounter` under a single `DispatchCounterMode`, inheriting all their correctness guarantees without reimplementing any dispatch logic.

**Horowitz presets** — `fp32()`, `fp16()`, `int8()` classmethods on `ComputeEnergyCostConfig` directly encode the canonical 45nm reference costs. The default matches FP32, making comparisons default to the most common regime in the SNN literature.

**Dual API surface** — Context manager (`ComputeEnergyProfiler`) + convenience function (`estimate_compute_energy`) follows the same pattern as `analytical_energy` and `spikesim` profilers. Consistent UX.

**Test coverage is solid** — 7 focused tests covering:
- Pure AC operations (element-wise add)
- Sparse linear (synop as auxiliary, AC counted)
- Dense linear (MAC only)
- Profiler ↔ convenience function equivalence
- Custom cost config application
- Preset value verification against reference table
- Default config matching FP32 preset
- FP16 cost regime switching

**Clean ancillary changes** — The PR correctly fixes:
- `:type:` annotation format in `speechcommands.py` (`type, optional` → `Optional[type]`)
- Typo "Defaule" → "Default" in `rnn.py`
- Missing bilingual docstrings on `rnn.py` functions, `memory_residency.py` methods, `neuron_state.py`, and `_sparse_memory.py`
- Sphinx/RST formatting in `misc.py`

---

### 💡 Suggestions (non-blocking)

#### 1. `get_counts()` return format — consider flat structure (Design)

**`compute_energy.py:279-281`** — Currently returns `{"counts": report.counts}`:

```python
def get_counts(self) -> dict[str, Any]:
    report = self.get_report()
    return {"counts": report.counts}
```

The existing profilers (`SpikeSimEventEnergyProfiler`, `NeuroMCEngine`) return flat dicts from `get_counts()` (no wrapping key). For API consistency, consider returning `report.counts` directly. Since there's only one key, the wrapping adds indirection with no benefit.

```python
def get_counts(self) -> dict[str, int]:
    return self.get_report().counts
```

#### 2. `warnings` field is never populated (Completeness)

**`compute_energy.py:273`** — `ComputeEnergyReport.warnings: list[str]` exists but is always set to `[]`. Consider one of:
- Remove the field until there's a concrete warning condition to report (cleaner API), or
- Add detection logic — e.g., warn when the model has `nn.Linear` but the input is non-spike (the MAC/AC boundary may be unintuitive without spike input), or when `extra_ignore_modules` cause partially missing coverage.

As-is, consumers who check `report.warnings` will never see one, which is misleading.

#### 3. Consider adding a Conv2d test (Test Quality)

**`test_op_counter_compute_energy.py`** — All tests use `nn.Linear` or element-wise `AddModel`. While the underlying `MACCounter`/`ACCounter` are tested separately with Conv2d, adding one test showing `estimate_compute_energy` works with a convolution would increase confidence in the composition. It can be short:

```python
def test_compute_energy_conv_counts_mac():
    model = nn.Conv2d(3, 6, 3, bias=False)
    x = torch.randn(1, 3, 32, 32)
    report = op_counter.estimate_compute_energy(model, x)
    assert report.counts["mac"] > 0
    assert report.counts["ac"] == 0
```

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review this PR\n@gemini-code-assist review this PR

**Comment by AllenYolk** [comment]
@coderabbitai review this PR\n@gemini-code-assist review this PR

File: spikingjelly/activation_based/op_counter/compute_energy.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    exact model of real kernels, mixed-precision accumulation paths, or a
+    specific hardware microarchitecture.
+    """
+
+    energy_total_pj: float
+    energy_mac_pj: float
+    energy_ac_pj: float
+    breakdown_pj: dict[str, float]
+    counts: dict[str, int]
+    warnings: list[str]
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/compute_energy.py:192]: 💡 **Suggestion:** The `warnings` field in `ComputeEnergyReport` is always initialized as `[]` in `get_report()` and never populated. Consider surfacing meaningful warnings here — e.g., when counters encounter unrecognized aten ops with `strict=False`, or when energy is zero because all counts are zero. This would make the field actionable rather than a forward-looking placeholder.

File: spikingjelly/activation_based/op_counter/compute_energy.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        self.ac_counter.reset()
+        self.synop_counter.reset()
+        self.flop_counter.reset()
+        self._dispatch_mode.__enter__()
+        return self
+
+    def __exit__(self, exc_type, exc, tb):
+        return self._dispatch_mode.__exit__(exc_type, exc, tb)
+
+    def get_report(self) -> ComputeEnergyReport:
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/compute_energy.py:253]: 💡 **Suggestion:** `get_report()` is the primary public API for retrieving results, but it has no docstring. Consider adding a bilingual docstring consistent with the project conventions (AGENTS.md requires docstrings for all public API). Describe the return type and the calculation formula.

File: spikingjelly/activation_based/op_counter/compute_energy.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+            counts={
+                "mac": mac,
+                "ac": ac,
+                "synop": synop,
+                "flop": flop,
+            },
+            warnings=[],
+        )
+
+    def get_total(self) -> float:
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/compute_energy.py:297]: 💡 **Suggestion:** Missing docstring for `get_total()`. Since this class is not a subclass of `BaseCounter` (which documents its `get_total()`), a standalone docstring would help users understand what this returns.

File: spikingjelly/activation_based/op_counter/compute_energy.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        return self.get_report().energy_total_pj
+
+    def get_counts(self) -> dict[str, int]:
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/compute_energy.py:300]: 💡 **Suggestion:** Missing docstring for `get_counts()`. Should document the dictionary keys (`mac`, `ac`, `synop`, `flop`) and the fact that `synop` and `flop` are auxiliary (not used in energy computation).

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Comment** — no blocking issues found. Clean implementation overall.

### ✅ Looks Good

| Aspect | Assessment |
|---|---|
| **Design** | Clean separation of concerns (cost config, profiler config, report, profiler class, convenience function). Reuses existing `MACCounter`, `ACCounter`, `SynOpCounter`, `FlopCounter` via `DispatchCounterMode` — good composition over reimplementation. |
| **Consistency** | Context manager API matches the existing profiler pattern (`LemaireEnergyProfiler`, etc.). The `estimate_compute_energy()` convenience function follows the same convention. |
| **Docstrings** | All classes and the public function have bilingual (CN/EN) Sphinx docstrings with proper RST reference labels — consistent with project standards. |
| **Tests** | 7 well-named test cases covering the core API surface: additive-only (AC), sparse linear (SynOp as auxiliary), dense (MAC), custom cost config, preset values, default-vs-fp32 equivalence, profiler-vs-convenience consistency. |
| **Imports** | `__all__` in `compute_energy.py` exports 5 names, no conflicts with existing exports. Single-line addition to `__init__.py`. |
| **RST docs** | Automodule section correctly placed after SynOp and before NeuroMC. |

### 💡 Suggestions (non-blocking)

1. **`warnings` field is never populated** (`compute_energy.py:187`) — The `ComputeEnergyReport` has a `warnings: list[str]` field that's always `[]`. Consider populating it with meaningful diagnostics (e.g., unrecognized aten ops when `strict=False`, zero-energy results). See inline comment.

2. **Missing docstrings on public API methods** — `get_report()`, `get_total()`, `get_counts()` on `ComputeEnergyProfiler` lack docstrings. Per project conventions (AGENTS.md), all public API should have bilingual docstrings. See inline comments.

3. **Consider adding grouped convolution tests** — The current tests only use `nn.Linear` and a pointwise-add model. A `nn.Conv2d(..., groups=N)` test would verify the group-division handling in `MACCounter`/`ACCounter` is correct when composed through `ComputeEnergyProfiler`.

4. **`extra_ignore_modules` is untested** — This config option is wired through to all four internal counters but has no test coverage.

### 📊 Changed Files (actual PR diff)

| File | Status | Lines |
|---|---|---|
| `spikingjelly/activation_based/op_counter/compute_energy.py` | **New** | 329 |
| `test/activation_based/test_op_counter_compute_energy.py` | **New** | 111 |
| `spikingjelly/activation_based/op_counter/__init__.py` | Modified | +1 |
| `docs/source/APIs/spikingjelly.activation_based.op_counter.rst` | Modified | +8 |

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review this PR\n@gemini-code-assist review this PR

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Reviewed 💬** (0 critical, 1 warning, 2 suggestions)

**PR:** #681 — Add a compute-only MAC/AC energy estimator with Horowitz presets
**Author:** @AllenYolk
**Files changed:** 13 (+541 -48)

### ⚠️ Warnings

#### (Documentation) — Missing `:type:`, `:return:`, `:rtype:` in public API docstring

`estimate_compute_energy()` is a public convenience function, but its docstring omits `:type:`, `:return:`, and `:rtype:` fields. Per the project's documentation conventions (AGENTS.md: "参数与类型：逐项写 `:param:` + `:type:`"), these should be present:

```python
def estimate_compute_energy(
    model: nn.Module,
    inputs,
    *,
    config: ComputeEnergyConfig | None = None,
) -> ComputeEnergyReport:
```

The docstring has `:param:`, but the params `model`, `inputs`, `config` are missing `:type:`, and there's no `:return:` or `:rtype:`. Same issue applies to `ComputeEnergyConfig`, `ComputeEnergyProfiler`, and `ComputeEnergyReport` class docstrings — they have `:param:` but no `:type:`.

### 💡 Suggestions

#### (Documentation) — Public methods missing docstrings

`ComputeEnergyProfiler.get_report()`, `get_total()`, and `get_counts()` are public methods with no docstrings. Even brief one-liners would help users of the API.

#### (Maintainability) — `_call_model` missing type annotation on `inputs`

The internal helper `_call_model` accepts `inputs` without a type annotation. Since it handles three dispatch paths (Tensor, tuple, list), annotating the parameter would clarify intent:

```python
def _call_model(model: nn.Module, inputs: torch.Tensor | tuple | list) -> torch.Tensor:
```

### ✅ Looks Good

- **Clean composition pattern** — `ComputeEnergyProfiler` composes four existing counters (`MACCounter`, `ACCounter`, `SynOpCounter`, `FlopCounter`) via `DispatchCounterMode`. This reuses proven code without reinventing dispatch logic.
- **Context manager + convenience function** — Both `with profiler:` and `estimate_compute_energy()` APIs are provided, matching the existing `LemaireEnergyProfiler` pattern.
- **Horowitz presets** — `fp32()`, `fp16()`, `int8()` classmethods on `ComputeEnergyCostConfig` are well-documented with reference values.
- **Type correctness** — `extra_ignore_modules: list[type[nn.Module]] | None = None` in `ComputeEnergyConfig` correctly uses types (matching the runtime `isinstance(am, tuple(counter.ignore_modules))` check in `BaseCounter` at base.py:708).
- **Test coverage** — 8 tests cover: pure AC path, sparse linear (SynOp), dense linear (MAC), profiler vs. convenience function consistency, custom cost config, all three presets (fp32/fp16/int8), default config matching fp32, and fp16 comparison regime.
- **No regressions** — All 21 existing `test_op_counter_snn_energy.py` tests continue to pass.
- **Export hygiene** — No name collisions in `__init__.py` `*` imports; all 5 new names are unique across the package.
- **RST correctness** — The new "Compute-Only Energy Estimator" section has a correct underline length (29 = 29). (Pre-existing underline issues in the file are outside this PR's scope.)

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review this PR\n@gemini-code-assist review this PR

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Reviewed 💬** (0 critical, 1 warning, 4 suggestions)

**PR:** #681 — Add a compute-only MAC/AC energy estimator with Horowitz presets
**Author:** @AllenYolk
**Files changed:** 13 (+1032 −28)

---

### Overview

This is a well-structured PR that adds a compute-only MAC/AC energy profiler (`ComputeEnergyProfiler`) composed from existing counters (`MACCounter`, `ACCounter`, `SynOpCounter`, `FlopCounter`) via `DispatchCounterMode`. The API is consistent with the project's existing profiler conventions (context manager + convenience function).

Horowitz 2014 cost presets (FP32/FP16/INT8) are provided with correct literature values. Tests cover the main usage scenarios with correct mathematical expectations.

The PR also includes a significant amount of docstring cleanup — adding bilingual documentation to existing modules (`_sparse_memory.py`, `memory_residency.py`, `neuron_state.py`, `rnn.py`) and fixing type annotations (`speechcommands.py`) and typos (`rnn.py`: "Defaule" → "Default"). These are welcome improvements.

---

### ⚠️ Warnings

#### 1. `strict` parameter in `ComputeEnergyConfig` is unused (Design / Documentation)

(Files: `compute_energy.py:132`, `compute_energy.py:222-238`)

`ComputeEnergyConfig` declares `strict: bool = False` with docstring stating it "only applies to profiler-level validation added by this wrapper." However, `strict` is never read in `ComputeEnergyProfiler.__init__()` or any other method — only `verbose` and `cost_config` are used.

Since the internal `DispatchCounterMode` is hard-coded as `strict=False` on line 236, there is no path where user-set `strict=True` propagates to the dispatcher. This is misleading: a user passing `ComputeEnergyConfig(strict=True)` expecting validation will get none.

**Fix:** Either (a) remove the field and update the docstring, or (b) thread it through to `DispatchCounterMode(strict=self.config.strict)`. Option (b) is cleaner since it provides actual value.

---

### 💡 Suggestions

#### 2. `warnings` field in `ComputeEnergyReport` is never populated

(Files: `compute_energy.py:190`, `compute_energy.py:276`)

The `ComputeEnergyReport` dataclass declares `warnings: list[str]` as a field, but `get_report()` always sets it to an empty list. Consider populating it with meaningful diagnostics, such as:
- Zero total count (no operations tracked — model may not contain supported layers)
- Mixed spike/dense warnings
- Unsupported module types encountered

#### 3. `_call_model` only handles `tuple/list` and single input

(Files: `compute_energy.py:24-27`)

```python
def _call_model(model: nn.Module, inputs):
    if isinstance(inputs, (tuple, list)):
        return model(*inputs)
    return model(inputs)
```

Models that accept keyword arguments (e.g., `attention_mask`, `input_ids`) are not supported, since `inputs` must be passed as a single positional argument or unpacked from a tuple/list. Consider adding `dict` support:
```python
if isinstance(inputs, dict):
    return model(**inputs)
```

The docstring already documents "tuple/list will be passed as `model(*inputs)`" but doesn't document the dict limitation.

#### 4. Consider adding a grouped convolution test

Tests cover linear layers and element-wise ops well. Since `MACCounter`/`ACCounter` have grouped-convolution rules (relevant to SNN workflows), a test with `nn.Conv2d(..., groups=N)` with known spike input would ensure the estimator composes correctly with grouped layers. This is especially relevant given the grouped-convolution formula subtleties documented in the project.

#### 5. Pre-existing RST underline mismatch in docs

(Files: `docs/source/APIs/spikingjelly.activation_based.op_counter.rst:5-6`)

Not introduced by this PR, but the section heading "Base Classes and Context Managers" (33 chars) has an underline 1 character too short (32 `+` signs). This would cause a Sphinx warning during `make html`. The new section added by this PR ("Compute-Only Energy Estimator") has correct underline length.

---

### ✅ Looks Good

- **Clean API design** — `ComputeEnergyProfiler` follows the established context-manager pattern, consistent with other profilers in the project. The `estimate_compute_energy()` convenience function is a natural entry point.

- **Correct energy formula** — `total_pj = mac * e_mac_pj + ac * e_ac_pj` is the right model for compute-only estimation.

- **Horowitz presets match literature** — FP32 (MAC=4.6pJ, AC=0.9pJ), FP16 (MAC=1.5pJ, AC=0.4pJ), INT8 (MAC=0.23pJ, AC=0.03pJ) are correct reference values from Horowitz 2014 (ISSCC).

- **Test coverage** — 8 tests covering pure-AC, spike-linear (AC+SynOp), dense-linear (MAC), profiler/convenience consistency, custom cost config, preset validation, and FP16 regime. All mathematical expectations verified against counter implementations.

- **Correct type annotation fix** — `extra_ignore_modules` type changed from `list[nn.Module]` (instances) to `list[type[nn.Module]]` (classes) in `memory_residency.py`. This is the correct type since `isinstance(module, tuple(ignore_modules))` requires class types.

- **Docstring improvements** — Consistent bilingual pattern across existing modules. Typo fixes ("Defaule" → "Default") and correct type annotation migration in `speechcommands.py`.

- **`__exit__` delegates to `DispatchCounterMode.__exit__`** which returns `None` (falsy), so exceptions propagate correctly — no accidental suppression.

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review this PR\n@gemini-code-assist review this PR

File: spikingjelly/activation_based/op_counter/compute_energy.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+__all__ = [
+    "ComputeEnergyCostConfig",
+    "ComputeEnergyConfig",
+    "ComputeEnergyProfiler",
+    "ComputeEnergyReport",
+    "estimate_compute_energy",
+]
+
+
+def _call_model(model: nn.Module, inputs):
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/compute_energy.py:24]: 💡 **Suggestion:** `inputs` parameter lacks a type annotation. Consider adding `torch.Tensor | tuple | list` (or similar) for consistency with the rest of the typed codebase.

**Review by AllenYolk — COMMENTED** [review]
## Code Review Summary

**Verdict: Reviewed 💬** (0 critical, 0 warnings, 4 suggestions)

**PR:** #681 — Add a compute-only MAC/AC energy estimator with Horowitz presets
**Author:** @AllenYolk
**Files changed:** 13 (+1064 -28)

---

### ✅ Looks Good

- **Clean architecture.** `ComputeEnergyProfiler` composes existing counters (`MACCounter`, `ACCounter`, `SynOpCounter`, `FlopCounter`) via `DispatchCounterMode` — nice reuse of tested components.
- **Well-documented presets.** Horowitz 2014 FP32/FP16/INT8 costs match reference table values (verified via tests). The default `cost_config` correctly defaults to FP32, and `fp32()`/`fp16()`/`int8()` classmethods are clean factory presets.
- **Dual API.** Both context manager (`ComputeEnergyProfiler`) and convenience function (`estimate_compute_energy()`) follow existing project patterns — consistent user experience.
- **Good edge case handling.** The zero-count check with `strict` mode toggle is well-designed — warns by default, raises on demand.
- **No regressions.** The formatting/docstring changes in `rnn.py`, `memory_residency.py`, `neuromc/core.py`, `speechcommands.py`, `_sparse_memory.py`, and existing test file are purely cosmetic/linguistic — no behavioral impact.
- **No `__all__` collisions.** Verify that the new `from .compute_energy import *` in `__init__.py` produces no name conflicts with existing modules.
- **Safe type hints.** Python >=3.11 (`pyproject.toml`) makes PEP 604 union syntax (`list[type[nn.Module]] | None`) fully native — no compatibility concern.

---

### 💡 Suggestions

- **`compute_energy.py:24`** — `_call_model`'s `inputs` parameter lacks a type annotation. Consider `torch.Tensor | tuple | list` for consistency with the rest of the typed codebase. (Maintainability)
- **`compute_energy.py:301`** — Same: `estimate_compute_energy`'s `inputs` parameter has no type hint. (Maintainability)
- **`compute_energy.py:24`** — `_call_model` hard-codes `tuple/list` unpacking. If the project later supports transformer-style models (which pass `**kwargs` via dicts), this won't dispatch correctly. Consider a future-proof dispatch or document as a known limitation in the docstring. (Completeness)
- **`test_op_counter_compute_energy.py`** — Test coverage is solid for core math but misses a few paths: (Test Quality)
  - No `Conv2d` test (grouped conv is common in SNN architectures — consider testing a grouped-vs-ungrouped scenario to validate the fanout formula).
  - `get_total()` and `get_counts()` shortcut methods on `ComputeEnergyProfiler` are untested — cheap to add.
  - `extra_ignore_modules` configuration exercised by no test.

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review this PR\n@gemini-code-assist review this PR

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: COMMENT** (Draft PR — non-blocking feedback)

Reviewed by Hermes Agent.

### P2 — Missing `:return:` and `:rtype:` in `estimate_compute_energy` docstring (Documentation)

The `estimate_compute_energy()` function has complete `:param:` / `:type:` fields in both Chinese and English sections of its docstring, but is missing `:return:` and `:rtype:` in both. This will cause incomplete Sphinx documentation output.

**Fix:** Add `:return:` / `:rtype:` after `:config:` in both language sections.

### P2 — No test for INT8 preset applied to a model (Test Coverage)

`test_compute_energy_cost_config_presets_match_horowitz_reference_table()` checks the preset values numerically, but there's no test applying `int8()` cost config to an actual model forward pass. The `fp16()` preset has an end-to-end test (`test_compute_energy_fp16_preset_changes_only_comparison_regime`), but `int8()` doesn't — a reader might reasonably wonder whether the INT8 preset is fully wired.

### P2 — INT8 preset E_MAC=0.23pJ seems slightly off from Horowitz table (Correctness)

Horowitz 2014 (Table II, 45nm) reports Mult8 = 0.2 pJ and Add8 = 0.03 pJ. E_MAC = 0.2 + 0.03 = 0.23 pJ. However, note that for INT8 the MAC operation cost should be `Mult8 + Add8 = 0.2 + 0.03 = 0.23 pJ` — this matches. The AC (addition-only) cost should be `Add8 = 0.03 pJ` — this also matches. No issue here after double-checking. ✅

### P3 — `_warnings.clear()` in `__enter__` may miss repeated-use semantics (Correctness)

If a user calls `get_report()` multiple times within the same `with` block, `_warnings` is not cleared between calls (it's cleared in `__enter__`, not `get_report`). The `matched_supported_ops` guard prevents double-appending the same warning, but this is implicit. Consider clearing warnings inside `get_report()` if the method is intended to be idempotent.

### P3 — `extra_ignore_modules` untested (Test Coverage)

The `extra_ignore_modules` parameter in `ComputeEnergyConfig` is wired through to all sub-counters but has no test coverage.

### ✅ What Was Done Well
- Clean API design with clear separation between cost config, profiler config, and report dataclass
- Good consistency test between context-manager and convenience function APIs (`test_compute_energy_profiler_matches_convenience_function`)
- Comprehensive edge case coverage: empty input with strict mode, unsupported ops warning/error, custom costs
- Bilingual docstring additions to `_sparse_memory.py`, `memory_residency.py`, `neuron_state.py`, `rnn.py` are thorough and correctly formatted
- Type annotation fix in `memory_residency.py` (`list[nn.Module]` → `list[type[nn.Module]]`)
- Docstring fixes (`Defaule` → `Default`, `type, optional` → `Optional[type]`) improve documentation quality

---

*Reviewed by Hermes Agent*

File: spikingjelly/activation_based/op_counter/compute_energy.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+    Convenience entry for compute-only MAC/AC energy estimation.
+    It runs one real forward pass and returns the energy report.
+
+    The default comparison regime is Horowitz 2014 FP32. For FP16 or INT8
+    comparisons, pass an explicit preset cost configuration.
+
+    :param model: model to profile
+    :param inputs: model input; tuple/list will be passed as ``model(*inputs)``
+    :param config: compute-only energy configuration
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/compute_energy.py:345]: **Missing `:return:` and `:rtype:` in docstring (English section)** — The `:param config:` ends the parameter block, but there's no `:return:` / `:rtype:` to document the return type (``ComputeEnergyReport``). Same issue in the Chinese section. This will produce incomplete Sphinx output.

File: spikingjelly/activation_based/op_counter/compute_energy.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    """
+    profiler = ComputeEnergyProfiler(config=config)
+    with profiler:
+        _ = _call_model(model, inputs)
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/compute_energy.py:349]: **Minor:** The model output is discarded via `_ = ...`. Consider just `model(*inputs)` / `model(inputs)` without assignment, since the output is unused.

**Comment by AllenYolk** [comment]
@coderabbitai review this PR\n@gemini-code-assist review this PR

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Reviewed 💬** (1 suggestion)

**PR:** #681 — Add a compute-only MAC/AC energy estimator with Horowitz presets
**Author:** @AllenYolk
**Files changed:** 13 (+1094 -30)

---

### 💡 Suggestions

- **compute_energy.py:310-346** (Documentation) — `estimate_compute_energy` docstring is comprehensive and bilingual, but lacks structured `:return:` and `:rtype:` Sphinx fields. The English section says "returns the energy report" verbally but doesn't use the project's required directives. Per the project conventions, public functions should have complete `:return:` / `:rtype:` (and `:raises:` where applicable) alongside the prose description.

- **compute_energy.py:250-251** (Maintainability) — `ComputeEnergyProfiler.__exit__` propagates the return value of `self._dispatch_mode.__exit__(...)`. While `DispatchCounterMode.__exit__` currently returns `None` (from `TorchDispatchMode.__exit__`), this return-forwarding pattern is fragile — if a future refactor of the parent `__exit__` ever returns `True` (suppressing exceptions), this wrapper would silently propagate exception suppression. Consider `return None` explicitly to be defensive.

---

### ✅ Looks Good

- **Clean architecture**: Clear separation of concerns between `ComputeEnergyCostConfig` (cost tables/presets), `ComputeEnergyConfig` (profiler behavior), `ComputeEnergyReport` (output data model), `ComputeEnergyProfiler` (context-manager API), and `estimate_compute_energy` (convenience function).

- **Comprehensive tests (10 test cases)**: Covers basic MAC/AC counting, sparse vs. dense linear layers, profiler-vs-convenience function consistency, custom cost configs, all three Horowitz presets, default-config matching FP32, warning/error behavior for unsupported ops, edge-case empty inputs, and dict/keyword-only model inputs.

- **Horowitz presets are correct**: Values match the Horowitz 2014 reference table (FP32 4.6/0.9 pJ, FP16 1.5/0.4 pJ, INT8 0.23/0.03 pJ).

- **Scope bookkeeping is correct**: The `matched_supported_ops` check uses `"Global"` scope which is correct — `ModuleTracker.__init__` initializes `self.parents = {"Global"}`, so `DispatchCounterMode` always records counts under both the module name and `"Global"`. This is consistent with `BaseCounter.get_total()` which also reads from `"Global"`.

- **Type annotation fix in `memory_residency.py:611`**: Changing `extra_ignore_modules` from `list[nn.Module]` to `list[type[nn.Module]]` is correct — `isinstance(am, tuple(counter.ignore_modules))` checks against module *classes*, not instances.

- **Docstring additions**: High-quality bilingual docstrings added to `memory_residency.py`, `neuron_state.py`, `rnn.py`, and `_sparse_memory.py` — consistent with project conventions.

- **Typo fixes**: "Defaule" → "Default" in `rnn.py`, RST backtick formatting fix in `misc.py`, type annotation standardization in `speechcommands.py`.

---

*Reviewed by Hermes Agent*

File: spikingjelly/activation_based/op_counter/compute_energy.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+
+    Convenience entry for compute-only MAC/AC energy estimation.
+    It runs one real forward pass and returns the energy report.
+
+    The default comparison regime is Horowitz 2014 FP32. For FP16 or INT8
+    comparisons, pass an explicit preset cost configuration.
+
+    :param model: model to profile
+    :param inputs: model input; tuple/list will be passed as ``model(*inputs)``
+    :param config: compute-only energy configuration
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/compute_energy.py:345]: 💡 **Suggestion:** Missing `:return:` and `:rtype:` directives. The prose says "returns the energy report" but doesn't follow the project's Sphinx RST convention. Add after the `:param config:` line:

```
:return: compute-only MAC/AC energy report
:rtype: ComputeEnergyReport
```
